### ISSUE TITLE
PSR12 Traits usage - invalid error

### ISSUE THREAD

**Issue by michalbundyra** [squizlabs/PHP_CodeSniffer#2691]
Consider the following example:

```php
<?php

namespace MyNamespace;

class MyExample
{
    use Trait1;
    protected $a = 1;
    use Trait2;
}
```

and run PSR12 checks on it:
```
----------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------------------------------------------------------------------------
 7 | ERROR | [x] There must be a blank line following the last trait import statement
   |       |     (PSR12.Traits.UseDeclaration.NoBlankLineAfterUse)
----------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------------------------------------------------------------------------
```

it is wrong, as it is not the last trait usage in the file.

Related to #2690

**Comment by gsherwood** [comment]
Yes, these 2 issues are related.

While PSR12 doesn't say that USE statements must be grouped together, it shows that in the sample code. I've already to change this sniff though as comments needs to be allowed between USE statements for documentation  purposes, so the sample code isn't quite correct.

While this sniff could go through the entire class looking for traits, it tries to shortcut this process and end when it finds the first non-use line. Performance will suffer if it needs to scan through the entire classes's token stack just in case someone has written a USE statement half way down the class. I'm not sure if it's worth it.

A possible compromise is to stop looking once a function keyword is found. It's not going to catch everything but it would catch this sample code, and there are likely far less tokens in the class properties than the methods.

What do you think?

**Comment by michalbundyra** [comment]
I am not sure what was the intention of PSR-12 about traits, but I believe they wanted all traits grouped, but this is just my guess. 

I don't know how big impact it's gonna have on performance if we check whole class for traits. We are looking just for `use` keywords and then we are checking if it is a trait or `use` in closure. Shouldn't be that bad.

Maybe there is a way to clarify it with PSR-12 if they meant all traits on top or not? Maybe they will add "errata" as they did for PSR-2? ;-)

**Comment by gsherwood** [comment]
> I don't know how big impact it's gonna have on performance if we check whole class for traits. We are looking just for `use` keywords and then we are checking if it is a trait or `use` in closure. Shouldn't be that bad.

It might not be actually. I was thinking about use findNext(), but this is faster done by doing the looping in the sniff itself as it can skip directly over methods.

> Maybe there is a way to clarify it with PSR-12 if they meant all traits on top or not? Maybe they will add "errata" as they did for PSR-2? ;-)

I wouldn't even bother trying :)

**Comment by gsherwood** [comment]
I'm thinking the best thing to do here might just be to implement the grouping check and see if anyone complains.

**Comment by michalbundyra** [comment]
I doubt people will start complaining about it ;-)

**Comment by michalbundyra** [comment]
@gsherwood I have tested changes and it seems to be fine for me. Thanks 👍

**Comment by gsherwood** [comment]
Thanks for testing.
